src/ophys_etl/decrosstalk/active_traces.py

```diff
@@ -0,0 +1,465 @@
+                          N : number of neurons (rois),
+                          M: number of timestamps (Fluorescence traces)
+
+    :param event_kernel: int, number of consecutive events
+
+    :param robust_std: boolean indicating whether or not
+                       to use the interquartile
+                       range to calcluate standard deviation
+
+    :return
+
+        idx_components: numpy.array;
+                        the components ordered according to the fitness
+
+        fitness: numpy.array;
+
+        erfc: numpy.array;
+              probability at each time step of observing the
+              N consecutive actual trace values given the distribution
+              of noise. For each event, this is the sum of the log
+              probabilities of finding a series of {event_kernel} events
+              at least as extreme as the {event_kernel} sequence of
+              events starting with designated event.
+
+    """
+
+    n_timesteps = np.shape(traces)[-1]
+    mode = mode_robust(traces, axis=1)
+    ff1 = traces - mode[:, None]
+
+    # only consider values under the mode
+    # to determine the noise standard deviation
+    ff1 = -ff1 * (ff1 < 0)
+
+    if robust_std:
+        # compute 25 percentile
+        ff1 = np.sort(ff1, axis=1)
+        ff1[ff1 == 0] = np.nan
+        ns = np.round(np.sum(ff1 > 0, 1) * .5).astype(int)
+        iqr_h = np.zeros(traces.shape[0])
+
+        for idx, el in enumerate(ff1):
+            iqr_h[idx] = ff1[idx, -ns[idx]]
+
+        # approximate standard deviation as iqr/1.349
+        sd_r = 2 * iqr_h / 1.349
+
+    else:
+        ns = np.sum(ff1 > 0, 1)
+        sd_r = np.sqrt(np.sum(ff1 ** 2, 1) / ns)
+
+    # compute z value
+    z = (traces - mode[:, None]) / (3 * sd_r[:, None])
+
+    # probability of observing values larger or equal to z given normal
+    # distribution with mean=mode and std sd_r
+    erf = 1 - scipy.stats.norm.cdf(z)
+
+    # use logarithm so that multiplication becomes sum
+    # np.errstate is to suppress "Divide by zero" warning from np.log
+    with np.errstate(divide='ignore'):
+        erf = np.log(erf)
+        # ToDo: Debug "RuntimeWarning: divide by zero encountered in log"
+
+    # build kernel for event detection
+    filt = np.ones(event_kernel)
+
+    # convolve probability with kernel
+    erfc = np.apply_along_axis(lambda m: np.convolve(
+        m, filt, mode='full'), axis=1, arr=erf)
+    erfc = erfc[:, :n_timesteps]
+
+    # select the maximum value of such probability for each trace
+    fitness = np.min(erfc, 1)
+
+    ordered = np.argsort(fitness)
+
+    idx_components = ordered  # [::-1]# selec only portion of components
+
+    #    fitness = fitness[idx_components] % FN
+    #    commented bc we want the indexing to match C and YrA.
+
+    #    erfc = erfc[idx_components] % FN
+    #    commented bc we want the indexing to match C and YrA.
+
+    return idx_components, fitness, erfc
+
+
+def find_event_gaps(evs):
```

Comment:
no type hinting, mixing docstring formats

Reply:
same for later functions

Reply:
I've updated the docstrings of the methods I've touched.

The others were copied from what I'm told is finely tuned code from scientific experts. I'm loath to touch those, since I haven't really altered the behavior of that code and want to preserve the original authors' intent as much as possible.

Reply:
ok